csrc/preseg_passes/allocation_order_inference.cpp

```diff
@@ -95,6 +96,12 @@ class AllocationOrderInferencer : public IterVisitor {
       const std::vector<IterDomain*>& permutation_ref) {
     if (auto iter = alloc_order_map_.find(producer);
         iter != alloc_order_map_.end()) {
+      // early termination to propagate empty allocation order
+      if (iter->second.empty()) {
+        alloc_order_map_[consumer] = {};
+        return true;
+      }
+
```

Comment:
Is this needed? If the following code is general enough to correctly handle rank-0 (which seems so to me), I'd rather remove the unnecessary if-then.

Reply:
This bit is necessary.

I think my PR comment wasn't clear vvv

> allocation_order_map_ allows empty allocation order for tensors created by factory method

So it's not really trying to handle rank-0 tensor, but instead it's trying to handle cases where a factory method produces a tensor, where we record it with an empty allocation order.

The line below:
```
      std::vector<IterDomain*> alloc_domain =
          propagateAllocationDomain(producer, consumer);
```
Would try to compute the alloc_domain from the allocation order.

(I mean, I can move this short-cut to `propagateAllocationDomain` if you think that looks cleaner.

Reply:
Can you remind me of the contract? Doesn't `alloc_order_map_[tv]` (if it exists in the map) need to have the same size as `tv`'s rfactor/allocation domain?

Reply:
^^^ That was the contract and that's what I changed in this PR.

I failed at my PR description.

There's three cases now, `alloc_order_map_[tv]` should either:

1. have the same size as the rank of `tv`'s rfactor domain;
    we have a preferred allocation order for the entry and it should participate in propagation.
2. it could be an empty array;
    it's a wild card, it shouldn't dictate output allocation order, but it still tells that propagation is not interrupted.
3. not exist in the map.
    this is the case where it says propagation has stopped in the branch. (likely due to lack of allocation order on inputs or operation not yet covered by propagation rule).

The reason is coming from two parts:
1. I don't want to blindly ignore other operands that doesn't have an allocation order when resolving multiple candidates: imagine we have something like vvv
```
nvfusion(tv0, tv1):
  tv2 = op0(tv0) # not covered by propagation
  tv3 = add(tv2, tv1)
  return tv3
```
If tv0 is a bigger tensor than tv1, ideally we should propagate output tv3 to be in the same memory format as in tv0. Because op0 is not covered by our propagation rule yet, we won't see the allocation order for tv2.
So when we try to resolve the propagation for operations that takes multiple operands, we can really only make the decision when we know for sure that every thing coming

2. tensors created by factory method should carry a wild-card and not affecting propagation. Because those tensors are not going to affect vectorization.

So with those two considerations, I decided to expand the alloc_order_map_ to allow an `empty` entry to disambiguate wild-card tensor from tensor where propagation has failed.